### ISSUE TITLE
About the input format is RGB image

### ISSUE THREAD

**Issue by PatrickWilliams44** [Thanos-DB/FullyConvolutionalTransformer#1]
Hi! Thank you very much for open source your code, if the input is three-channel RGB image, then how should I modify the code?

**Comment by Thanos-DB** [comment]
Hi Patrick,

thank you for your interest in our paper.
You did not tell us what you have already tried and what the error was. 
Please pass your data into the function to create the model the same way as we do and let us know if this does not work.

-Thanos

**Comment by PatrickWilliams44** [comment]
Hi Thanos-DB! Thank you very much for your timely reply, I have solved this problem now, this is a novel design model, thank you again for sharing the code!

**Comment by Thanos-DB** [comment]
I am glad it works now.

**Comment by PatrickWilliams44** [comment]
Hello! Thank you for your patient reply! I have run the FCT model on my own dataset, in which the trainset contains 800 images and the testset contains 200 images. In the training process, the loss gradually decreases and converges, but the performance on the testset is not satisfactory, and the dice coefficient is very low, even only 0.2. I don't know what causes this. Can you give me your suggestion?

I am looking forward to your reply and have best wishes for you!

Gratitude, beyond words!



On 10/27/2022 ***@***.***> wrote：

I am glad it works now.

**Comment by chaitanya-kaul** [comment]
Hi Patrick,

It is impossible to exactly point out the reason for this without any further information on the nuances of the problem you're trying to tackle.

Some possible reasons for this could be the following - 
- Class imbalance in pixels that you want to detect in the image vs background pixels
- The model is overparameterized for your problem and you may need to tweak it
- The distribution of the data in the training and testing set may not be similar
- The model converges at an earlier point but starts to overfit due to any overtraining
- Can be a few more..

We've tested FCT on 5 different datasets and have never encountered such a problem. This is because an issue such as the one you have, is highly data dependant rather than model dependant i.e. the model needs to be tweaked to make it work with the data.

**Comment by Thanos-DB** [comment]
Hi Patrick,

Chaitanya already replied with my thoughts exactly.
It could be caused by a number of things. Is your data publicly available? Have you tested any other models? Can you try to overfit the model (i.e. use very few images not all 800 to try to force it to overfit and see if this is possible at all). Do you have any plots maybe using TensorBoard during training (for this one you also need a validation set, only having the training set will not help)? Before training also check that the generators are set correctly (for instance if you rotate your images see that you do the exact same thing with the masks so plot X and y using the output of the generator). Without knowing what you are doing (so i may be wrong here) 0.2 is too low so i assume it is an implementation issue or as Chaitanya pointed out it has to do with your data. In all of the datasets that we have used it was never this low unless there was a bug somewhere.

**Comment by PatrickWilliams44** [comment]
Hi Chaitanya,

Thank you very much for your patient explanation. I will continue the experiment according to the guidance method you provided me. At the same time, I will also check my own data set. I believe that the FCT model has good enough performance for medical image segmentation, so I will make full use of this model for academic research. Thank you again for your timely response. Thank you so much!

Best wishes to you!

On 11/1/2022 ***@***.***> wrote：

Hi Patrick,

It is impossible to exactly point out the reason for this without any further information on the nuances of the problem you're trying to tackle.

Some possible reasons for this could be the following -

Class imbalance in pixels that you want to detect in the image vs background pixels
The model is overparameterized for your problem and you may need to tweak it
The distribution of the data in the training and testing set may not be similar
The model converges at an earlier point but starts to overfit due to any overtraining
Can be a few more..

We've tested FCT on 5 different datasets and have never encountered such a problem. This is because an issue such as the one you have, is highly data dependant rather than model dependant i.e. the model needs to be tweaked to make it work with the data.

**Comment by PatrickWilliams44** [comment]
Hi Thanos-DB,

I have received the reply from Chaitanya, and thank you for your professional guidance, which will be very helpful for my further academic research. The performance indicators of the FCT model are excellent enough to be competent for most medical image segmentation tasks, so I will continue my research based on this model and fully explore new ideas and on this basis. Thank you again for your timely reply! Gratitude expressed in words.

 

Best wishes to you!



On 11/1/2022 ***@***.***> wrote：

Hi Patrick,

Chaitanya already replied with my thoughts exactly.
It could be caused by a number of things. Is your data publicly available? Have you tested any other models? Can you try to overfit the model (i.e. use very few images not all 800 to try to force it to overfit and see if this is possible at all). Do you have any plots maybe using TensorBoard during training (for this one you also need a validation set, only having the training set will not help)? Before training also check that the generators are set correctly (for instance if you rotate your images see that you do the exact same thing with the masks so plot X and y using the output of the generator). Without knowing what you are doing (so i may be wrong here) 0.2 is too low so i assume it is an implementation issue or as Chaitanya pointed out it has to do with your data. In all of the datasets that we have used it was never this low unless there was a bug somewhere.

**Comment by PatrickWilliams44** [comment]
Hi Thanos DB, 

I'm sorry to bother you again, but I still have a problem that I hope to get your help. After receiving the email reply from you and Chaitanya Kaul, I started to solve the problem according to the ideas you provided me.

So, I used the pytorch framework to conduct experiments on the Kvasir dataset with FCT model. For the specific implementation of the code, please refer to the attachment, where the trainset: validset: testset=8:1:1, and the size of the input image was adjusted to 224 × 224. For preprocessing operations such as horizontal flipping and vertical flipping of the trainset, BCELoss was used as the loss function, and the learning rate was 1e-3. Adam was used as the optimizer, and dice coefficient was used to monitor the validset. Unfortunately, dice coefficient was still very low.

In order to verify whether there was a problem in Kvasir dataset, I used other models to conduct experiments on Kvasir dataset, and the results showed that dice coefficient was normal when other models were used. This problem has been puzzling me all the time. I really want to find out where my problem occurred.

I have attached my specific experimental code about FCT to the attachment this time, so that you can check out some of my problems. At the same time, I'm sorry for taking up your time, and I'm grateful for your patient guidance. Thank you again! Gratitude beyond my words!

Best wishes to you!



On 11/1/2022 ***@***.***> wrote：

Hi Patrick,

Chaitanya already replied with my thoughts exactly.
It could be caused by a number of things. Is your data publicly available? Have you tested any other models? Can you try to overfit the model (i.e. use very few images not all 800 to try to force it to overfit and see if this is possible at all). Do you have any plots maybe using TensorBoard during training (for this one you also need a validation set, only having the training set will not help)? Before training also check that the generators are set correctly (for instance if you rotate your images see that you do the exact same thing with the masks so plot X and y using the output of the generator). Without knowing what you are doing (so i may be wrong here) 0.2 is too low so i assume it is an implementation issue or as Chaitanya pointed out it has to do with your data. In all of the datasets that we have used it was never this low unless there was a bug somewhere.

**Comment by Thanos-DB** [comment]
Hi Patrick,

i cannot see any attachment so i cannot really comment.
Is it still 0.2 or did you make any progress? What models did you use and what is the Dice coefficient for each model? Also, while training did you check your loss for the validation and test sets?

**Comment by PatrickWilliams44** [comment]
Hi Thanos DB, 
I tried to use the training set instead of the test set to observe whether the model learned something in the training process. The result was satisfactory, and the dice coefficient was indeed increasing gradually; Because at the beginning, I only used 5 epochs for training. When I tried to use more epochs for training, the dice coefficient was also rising steadily, which was gratifying! I'm very sorry to take up so much of your time, and I also express my high respect and sincere thanks for your patient reply! Gratitude is expressed in words!

Best wishes to you!


On 11/7/2022 ***@***.***> wrote：

Hi Patrick,

i cannot see any attachment so i cannot really comment.
Is it still 0.2 or did you make any progress? What models did you use and what is the Dice coefficient for each model? Also, while training did you check your loss for the validation and test sets?
